core/src/main/java/io/questdb/griffin/engine/functions/catalogue/WalTableListFunctionFactory.java

```diff
@@ -0,0 +1,266 @@
+                        boolean isWalTable = TableSequencerAPI.isWalTable(tableNameSink, rootPath, ff);
+                        rootPath.trimTo(rootLen);
+                        if (isWalTable) {
+                            if (record.switchTo(tableNameSink)) {
+                                return true;
+                            }
+                        }
+                    }
+                }
+            }
+
+            @Override
+            public void recordAt(Record record, long atRowId) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public long size() {
+                return -1;
+            }
+
+            @Override
+            public void toTop() {
+                close();
+            }
+
+            public class TableListRecord implements Record {
+                private long sequencerTxn;
+                private boolean suspendedFlag;
+                private long writerTxn;
+
+                @Override
+                public boolean getBool(int col) {
+                    if (col == suspendedColumn) {
+                        return suspendedFlag;
+                    }
+                    return false;
+                }
+
+                @Override
+                public long getLong(int col) {
+                    if (col == writerTxnColumn) {
+                        return writerTxn;
+                    }
+                    if (col == sequencerTxnColumn) {
+                        return sequencerTxn;
+                    }
+                    return Numbers.LONG_NaN;
+                }
+
+                @Override
+                public CharSequence getStr(int col) {
+                    if (col == nameColumn) {
+                        return tableNameSink;
+                    }
+                    return null;
+                }
+
+                @Override
+                public CharSequence getStrB(int col) {
+                    return getStr(col);
+                }
+
+                @Override
+                public int getStrLen(int col) {
+                    return getStr(col).length();
+                }
+
+                public boolean switchTo(final CharSequence tableName) {
+                    int rootLen = rootPath.length();
+                    rootPath.concat(tableName).concat(SEQ_DIR);
+                    long fdMeta = -1;
+                    long fdTxn = -1;
+                    try {
+                        fdMeta = TableUtils.openRO(ff, rootPath, META_FILE_NAME);
+                        fdTxn = TableUtils.openRO(ff, rootPath, TXNLOG_FILE_NAME);
+                        suspendedFlag = ff.readNonNegativeByte(fdMeta, SEQ_META_SUSPENDED) > 0;
+                        sequencerTxn = ff.readNonNegativeLong(fdTxn, MAX_TXN_OFFSET);
+                    } finally {
+                        rootPath.trimTo(rootLen);
+                        if (fdMeta > -1) {
+                            ff.close(fdMeta);
```

Comment:
this is a super-common thing, not to pick on you here, but we can do a good deed and introduce a new helper function on `ff`, which will close `fd` conditionally. there are about 90 other places that do the same

Reply:
Fixed in https://github.com/questdb/questdb/pull/2841/commits/510509c0dc6becef11c8450365228117911878f7